README.md
# NAME

AproJo - A time recording application based on Mojolicious

<div>
    <a href='https://travis-ci.org/wollmers/AproJo'><img src='https://travis-ci.org/wollmers/AproJo.png' alt="AproJo" /></a>
    <a href='https://coveralls.io/r/wollmers/AproJo?branch=master'><img src='https://coveralls.io/repos/wollmers/AproJo/badge.png?branch=master' alt='Coverage Status' /></a>
    <a href='http://cpants.cpanauthors.org/dist/AproJo'><img src='http://cpants.cpanauthors.org/dist/AproJo.png' alt='Kwalitee Score' /></a>
    <a href='http://badge.fury.io/pl/AproJo'><img src='https://badge.fury.io/pl/AproJo.svg' alt='CPAN version' height='18' /></a>
</div>

# SYNOPSIS

    $ aprojo setup
    $ aprojo daemon

# DESCRIPTION

[AproJo](https://metacpan.org/pod/AproJo) is a Perl web application.

# STATUS

[AproJo](https://metacpan.org/pod/AproJo) is still in pre-Alpha state. It still misses essential features to be useful in production.

# INSTALLATION

[AproJo](https://metacpan.org/pod/AproJo) uses well-tested and widely-used CPAN modules, so installation should be as simple as

    $ cpanm AproJo

when using [App::cpanminus](https://metacpan.org/pod/App::cpanminus). Of course you can use your favorite CPAN client or install manually by cloning the ["SOURCE REPOSITORY"](#source-repository).

# SETUP

## Environment

Although most of [AproJo](https://metacpan.org/pod/AproJo) is controlled by a configuration file, a few properties must be set before that file can be read. These properties are controlled by the following environment variables.

- `APROJO_HOME`

    This is the directory where [AproJo](https://metacpan.org/pod/AproJo) expects additional files. These include the configuration file and log files. The default value is the current working directory (`cwd`).

- `APROJO_CONFIG`

    This is the full path to a configuration file. The default is a file named `aprojo.conf` in the `APROJO_HOME` path, however this file need not actually exist, defaults may be used instead. This file need not be written by hand, it may be generated by the `aprojo config` command.

## The `aprojo` command line application

[AproJo](https://metacpan.org/pod/AproJo) installs a command line application, `aprojo`. It inherits from the [mojo](https://metacpan.org/pod/mojo) command, but it provides extra functions specifically for use with AproJo.

### config

    $ aprojo config [options]

This command writes a configuration file in your `APROJO_HOME` path. It uses the preset defaults for all values, except that it prompts for a secret. This can be any string, however stronger is better. You do not need to memorize it or remember it. This secret protects the cookies employed by AproJo from being tampered with on the client side.

[AproJo](https://metacpan.org/pod/AproJo) does not need to be configured, however it is recommended to do so to set your application's secret.

The `--force` option may be passed to overwrite any configuration file in the current working directory. The default is to die if such a configuration file is found.

### setup

    $ aprojo setup

This step is required. Run `aprojo setup` to setup a database. It will use the default DBI settings (SQLite) or whatever is setup in the `APROJO_CONFIG` configuration file.

# RUNNING THE APPLICATION

    $ aprojo daemon

After the database is has been setup, you can run `aprojo daemon` to start the server.

You may also use [morbo](https://metacpan.org/pod/morbo) (Mojolicious' development server) or [hypnotoad](https://metacpan.org/pod/hypnotoad) (Mojolicious' production server). You may even use any other server that Mojolicious supports, however for full functionality it must support websockets. When doing so you will need to know the full path to the `aprojo` application. A useful recipe might be

    $ hypnotoad `which aprojo`

where you may replace `hypnotoad` with your server of choice.

## Logging

Logging in [AproJo](https://metacpan.org/pod/AproJo) is the same as in [Mojolicious](https://metacpan.org/pod/Mojolicious::Lite#Logging). Messages will be printed to `STDERR` unless a directory named `log` exists in the `APROJO_HOME` path, in which case messages will be logged to a file in that directory.

# TECHNOLOGIES USED

- [Mojolicious](http://mojolicio.us) - a next generation web framework for the Perl programming language
- [DBIx::Class](http://www.dbix-class.org/) - an extensible and flexible Object/Relational Mapper written in Perl
- [Bootstrap](http://twitter.github.com/bootstrap) - the CSS/JS library from Twitter
- [jQuery](http://jquery.com/) - jQuery

# SEE ALSO

- [Contenticious](https://metacpan.org/pod/Contenticious) - File-based Markdown website application

# SOURCE REPOSITORY

[http://github.com/wollmers/AproJo](http://github.com/wollmers/AproJo)

# AUTHOR

Helmut Wollmersdorfer, <helmut.wollmersdorfer@gmail.com>

<div>
    <a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>
</div>

# COPYRIGHT AND LICENSE

Copyright (C) 2013-2015 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for AproJo.

{{$NEXT}}

0.019  2015-04-24 16:14:16 CEST
       - improve tests and coverage

0.018  2015-04-22 14:30:01 CEST
       - improved travis coverage

0.017  2015-04-22 11:56:57 CEST
       - improve author tests

0.016  2015-04-21 15:48:36 CEST
       - added spelling tests (AUTHOR_TESTING only)

0.015  2015-04-19 12:31:47 CEST
       - improve pod

0.014  2015-04-13 10:57:15 CEST
       - note on status

0.013  2015-04-08 13:54:55 CEST
       - push version of Plugin::Form

0.012  2015-04-07 14:31:48 CEST
       - debugging tests

0.011  2015-04-01 17:54:09 CEST
       - adapt dependencies

0.010  2015-03-31 15:31:46 CEST
       - adapt dependencies

0.009  2015-03-31 14:28:53 CEST
       - add travis
       - add badges
       - ServerInfo and DBInfo are now plugins with their own dist

0.008  2014-05-08 14:16:14 CEST
       - move Mojolicious::Plugin::Form to its own dist

0.007  2014-05-08 12:27:41 CEST
       - try a developer version

0.006  2014-05-07 15:27:39 CEST
       - adapted foldernames for CPAN

0.005  2014-05-02 12:50:08 CEST
       - correct versions

0.004  2014-04-28 17:13:14 CEST
       - debug failed login test

0.003  2014-04-24 12:20:09 CEST
       - debug warnings (fails on CPANTesters), cleanup

0.002  2014-04-23 15:21:26 CEST
       - need Perl v5.10.1 as minimum

0.001  2014-04-23 
       - initial version


lib/AproJo.pm
package AproJo;
use Mojo::Base 'Mojolicious';

use Data::Dumper;

our $VERSION = '0.019';

use File::Basename 'dirname';
use File::Spec;
use File::Spec::Functions qw'rel2abs catdir';
use File::ShareDir 'dist_dir';
use Cwd;

use Mojo::Home;

has db => sub {
  my $self         = shift;
  my $schema_class = $self->config->{db_schema}
    or die "Unknown DB Schema Class";
  eval "require $schema_class"
    or die "Could not load Schema Class ($schema_class), $@\n";

  my $db_connect = $self->config->{db_connect}
    or die "No DBI connection string provided";
  my @db_connect = ref $db_connect ? @$db_connect : ($db_connect);

  my $schema = $schema_class->connect(@db_connect)
    or die "Could not connect to $schema_class using $db_connect[0]";

  return $schema;
};

has app_debug => 0;

has home => sub {
  my $path = $ENV{MOJO_HOME} || getcwd;
  return Mojo::Home->new(File::Spec->rel2abs($path));
};

has config_file => sub {
  my $self = shift;
  return $ENV{APROJO_CONFIG} if $ENV{APROJO_CONFIG};

  return $self->home->rel_file('aprojo.conf');
};

sub startup {
  my $app = shift;

  $app->plugin(
    Config => {
      file    => $app->config_file,
      default => {
        'db_connect' => [
          'dbi:SQLite:dbname=' . $app->home->rel_file('aprojo.db'),
          undef,
          undef,
          {'sqlite_unicode' => 1}
        ],
        'db_schema' => 'AproJo::DB::Schema',
        'secret'    => '47110815'
      },
    }
  );

  {
    # use content from directories under share/files or using File::ShareDir
    my $lib_base = catdir(dirname(rel2abs(__FILE__)), '..', 'share','files');

    my $public = catdir($lib_base, 'public');
    $app->static->paths->[0] = -d $public ? $public : catdir(dist_dir('AproJo'), 'files','public');
    my $static_path = $app->static->paths->[0];
    #print STDERR '$static_path: ',$static_path,"\n";

    my $templates = catdir($lib_base, 'templates');
    $app->renderer->paths->[0] = -d $templates ? $templates : catdir(dist_dir('AproJo'), 'files', 'templates');
  }

  $app->plugin('I18N');
  $app->plugin('Mojolicious::Plugin::ServerInfo');
  $app->plugin('Mojolicious::Plugin::DBInfo');

  $app->plugin('Mojolicious::Plugin::Form');

  push @{$app->commands->namespaces}, 'AproJo::Command';

  #DEPRECATED: $app->secret( $app->config->{secret} );
  $app->secrets([$app->config->{secret}]);

  $app->helper(schema => sub { shift->app->db });

  $app->helper('home_page' => sub {'/'});

  $app->helper(
    'source_id' => sub {
      my ($self, $source) = @_;
      return undef unless $source;
      my @columns    = $self->schema->source($source)->columns;
      my $table_name = $self->schema->class($source)->table;
      my $source_id  = $table_name . '_id';
      return $source_id if (grep {/$source_id/} @columns);
      return $columns[0] if (scalar @columns);
    }
  );

  $app->helper(
    'get_user' => sub {
      my ($self, $name) = @_;
      unless ($name) {
        $name = $self->session->{username};
      }
      return undef unless $name;
      #say STDERR 'get_user: ', $name if $self->app->app_debug;
      return $self->schema->resultset('User')->single({name => $name});
    }
  );

  $app->helper(
    'has_role' => sub {
      my $self = shift;
      my $user_string = shift || '';
      my $role_string = shift || '';
      my $user = $self->get_user($user_string);
      return undef unless $user;
      my $role = $user->roles()->single({name => $role_string});
      #say STDERR 'has_role: ', $role->name if $self->app->app_debug;
      return ($role && $role->name eq $role_string);
    }
  );
  $app->helper(
    'is_admin' => sub {
      my ($self,$user) = @_;
      return $self->has_role($user,'admin');
    }
  );

  my $routes = $app->routes;

  $routes->get('/')->to('front#index');
  $routes->get('/front/*name')->to('front#page');
  $routes->post('/save')->to('front#save');
  $routes->post('/login')->to('user#login');
  $routes->any('/logout')->to('user#logout');

  my $if_admin = $routes->under(
    sub {
      my $self = shift;
      return $self->auth_fail unless $self->is_admin;
      return 1;
    }
  );

  $if_admin->post('/admin/save/:table')->to('admin#save');

  $if_admin->get('/admin/change/:table/:id')->to('admin#change');
  $if_admin->get('/admin/show/:table')->to('admin#show');

}

1;

__END__

=head1 NAME

AproJo - A time recording application based on Mojolicious

=begin html

<a href='https://travis-ci.org/wollmers/AproJo'><img src='https://travis-ci.org/wollmers/AproJo.png' alt="AproJo" /></a>
<a href='https://coveralls.io/r/wollmers/AproJo?branch=master'><img src='https://coveralls.io/repos/wollmers/AproJo/badge.png?branch=master' alt='Coverage Status' /></a>
<a href='http://cpants.cpanauthors.org/dist/AproJo'><img src='http://cpants.cpanauthors.org/dist/AproJo.png' alt='Kwalitee Score' /></a>
<a href='http://badge.fury.io/pl/AproJo'><img src='https://badge.fury.io/pl/AproJo.svg' alt='CPAN version' height='18' /></a>

=end html

=head1 SYNOPSIS

 $ aprojo setup
 $ aprojo daemon

=head1 DESCRIPTION

L<AproJo> is a Perl web application.

=head1 STATUS

L<AproJo> is still in pre-Alpha state. It still misses essential features to be useful in production.

=head1 INSTALLATION

L<AproJo> uses well-tested and widely-used CPAN modules, so installation should be as simple as

    $ cpanm AproJo

when using L<App::cpanminus>. Of course you can use your favorite CPAN client or install manually by cloning the L</"SOURCE REPOSITORY">.

=head1 SETUP

=head2 Environment

Although most of L<AproJo> is controlled by a configuration file, a few properties must be set before that file can be read. These properties are controlled by the following environment variables.

=over

=item C<APROJO_HOME>

This is the directory where L<AproJo> expects additional files. These include the configuration file and log files. The default value is the current working directory (C<cwd>).

=item C<APROJO_CONFIG>

This is the full path to a configuration file. The default is a file named F<aprojo.conf> in the C<APROJO_HOME> path, however this file need not actually exist, defaults may be used instead. This file need not be written by hand, it may be generated by the C<aprojo config> command.

=back

=head2 The F<aprojo> command line application

L<AproJo> installs a command line application, C<aprojo>. It inherits from the L<mojo> command, but it provides extra functions specifically for use with AproJo.

=head3 config

 $ aprojo config [options]

This command writes a configuration file in your C<APROJO_HOME> path. It uses the preset defaults for all values, except that it prompts for a secret. This can be any string, however stronger is better. You do not need to memorize it or remember it. This secret protects the cookies employed by AproJo from being tampered with on the client side.

L<AproJo> does not need to be configured, however it is recommended to do so to set your application's secret.

The C<--force> option may be passed to overwrite any configuration file in the current working directory. The default is to die if such a configuration file is found.

=head3 setup

 $ aprojo setup

This step is required. Run C<aprojo setup> to setup a database. It will use the default DBI settings (SQLite) or whatever is setup in the C<APROJO_CONFIG> configuration file.

=head1 RUNNING THE APPLICATION

 $ aprojo daemon

After the database is has been setup, you can run C<aprojo daemon> to start the server.

You may also use L<morbo> (Mojolicious' development server) or L<hypnotoad> (Mojolicious' production server). You may even use any other server that Mojolicious supports, however for full functionality it must support websockets. When doing so you will need to know the full path to the C<aprojo> application. A useful recipe might be

 $ hypnotoad `which aprojo`

where you may replace C<hypnotoad> with your server of choice.

=head2 Logging

Logging in L<AproJo> is the same as in L<Mojolicious|Mojolicious::Lite/Logging>. Messages will be printed to C<STDERR> unless a directory named F<log> exists in the C<APROJO_HOME> path, in which case messages will be logged to a file in that directory.

=head1 TECHNOLOGIES USED

=over

=item *

L<Mojolicious|http://mojolicio.us> - a next generation web framework for the Perl programming language

=item *

L<DBIx::Class|http://www.dbix-class.org/> - an extensible and flexible Object/Relational Mapper written in Perl

=item *

L<Bootstrap|http://twitter.github.com/bootstrap> - the CSS/JS library from Twitter

=item *

L<jQuery|http://jquery.com/> - jQuery


=back

=head1 SEE ALSO

=over

=item *

L<Contenticious> - File-based Markdown website application

=back

=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/AproJo>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=begin html

<a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>

=end html

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2015 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



      
